py/dml/ctree.py

```diff
@@ -3953,13 +3953,28 @@ def mkCast(site, expr, new_type):
         old_type = safe_realtype(expr.ctype())
         if real.cmp(old_type) == 0:
             return expr
-        else:
+        elif old_type.is_arith or isinstance(old_type, TPtr):
             return mkApply(
                 expr.site,
                 mkLit(expr.site, *real.get_store_fun()),
                 (mkCast(expr.site, expr, TInt(64, False)),))
+        else:
+            raise ECAST(site, expr, new_type)
+    if ((real.is_arith and old_type.is_arith)
+        or ((isinstance(real, TPtr) or real.is_int)
+            and (isinstance(old_type, (TPtr, TArray)) or old_type.is_int))):
+        assert (not (real.is_int and real.is_endian)
+                and not (old_type.is_int and old_type.is_endian))
+        return Cast(site, expr, new_type)
+
+    # Allow unsafe casts from method references to function pointers in DML 1.2
+    # for compatibility reasons
+    if (dml.globals.dml_version == (1, 2) and isinstance(expr, NodeRef)
+        and expr.get_ref()[0].objtype == 'method'
+        and isinstance(real, TPtr) and isinstance(real.base, TFunction)):
+        return Cast(site, expr, new_type)
```

Comment:
I wish to be rid of this if possible. We must test against VP to make sure that's safe to do, and `1.2/misc/destroy` will need to be rewritten.

Reply:
why bother? it's within a 1.2 guard, and reflects a (sad) 1.2 idiom. IIRC I have seen it used for hap callbacks. You could consider restricting the 1.2 guard to check site.file_info.dml_version instead.